five_tips_about_testable_golang_code.md

```diff
@@ -0,0 +1,171 @@
+# How to write UNtestable code in golang
+According to my experience in movies, if you want to be a super bad guy, you will need to work harder than the good guys. Think about it! You need to hire evil scientists, build a huge death bomb to threaten the world, a good plan and a lot of money. But be a good guy you just basically need some brave hearts, you do not even have to know about the structure of bomb because when choosing red/blue wire to cut, good guys never mess up.
+
+Yeah, that's not fair.
+
+Good news (?) everyone, in software developer side, mess up code is far easier than build it in a good way. If you keep following points in mind, make golang code untestable would be so easy.
+
+Before reading this, I assume you already read [Aaron's Blog: Testify - A Toolkit for Golang Unit Tests](https://vendasta-blog.appspot.com/blog/BL-F5QKDT6H/).
+
+## 1. Always expected golang mock as powerful as Python mock
+
+This is the easiest part. If you start with something with Python, all you need is ~~love~~ do nothing extra, just write code with all Python style, you can make it untestable.
+
+First, let take a look at this chart:
+
+|        | python mock | golang testify mock |
+|--------|-------------|-------------|
+|mock a class|  Yes|         Yes    |
+|mock a method in a mock class| Yes |Yes|
+|assert mock been called with given parameter | Yes | Yes|
+|return any result in mock method | Yes |Yes|
+|mock an attribute in a mock class | Yes | No |
+|mock a function | Yes | No|
+| patch | Yes | No |
+
+![that is how good Python is.](https://imgs.xkcd.com/comics/python.png)
+
+Mock(testify) in golang is good. Just not quite as *magical* as mocking in Python. When I first learn how to use it, the thought occurred to me most frequently was "What, I need to do this manually?" and "If I need to do this manually, what are YOU doing, mock?"
+
+~~" I am Groot." Mock replied.~~
+
+Mock in python can automatically do two things, one is **replace current real code to mock code**, another is **verify calls happen as expected**.
+
+Mock in go testify can do the second job(verify calls) but not first one(replace real code to mock). That part has to be done manually. And manually replacement have more limitations than magic replacement.
+
+We can talk about the limitations in next sections, for this section, just simply forget golang mock is not as powerful as python's mock, is a good start to write untestable code in go.
```

Comment:
just simply forgetting that mock is not as powerful as python's mock, is a good start to writing untestable code in go